public/js/mstream.js

```diff
@@ -383,6 +385,60 @@ $(document).ready(function () {
   // Stores an array of searchable objects
   var currentBrowsingList = [];
 
+  var savedExplorerListStatesPerScreen = {
+    fileExplorer: [],
+    allPlaylists: [],
+    allAlbums: [],
+    allArtists: [],
+    allRated: [],
+    recentlyAdded: []
+  }
+
+  function currentScreenExplorerListStates() {
+    if (programState.length > 0) {
+      return savedExplorerListStatesPerScreen[programState[0].state];
+    }
+  }
+
+  function saveExplorerListState() {
+    const statesList = currentScreenExplorerListStates();
+    if (statesList) {
+      statesList.push({
+        filelistScrollTop: $('#filelist').scrollTop(),
+        searchFoldersVal: $('#search_folders').val()
+      });
+    }
+  }
+
+  function saveExplorerListStateBeforeScreenChange() {
+    if (programState.length == 1) {
+      saveExplorerListState();
+    }
+  }
+
+  function restorePreviousExplorerListState() {
+    const statesList = currentScreenExplorerListStates();
+    if (statesList) {
+      const previousState = statesList.pop();
+      if (previousState) {
+        restoreExplorerListState(previousState);
+      }
+    }
+  }
+
+  function restoreExplorerListAfterScreenChange() {
+    const statesList = currentScreenExplorerListStates();
+    if (statesList.length > 0) {
+      restoreExplorerListState(statesList[0]);
+      savedExplorerListStatesPerScreen[programState[0].state] = [];
```

Comment:
we restore only state for root of a screen, maybe it would be better to clear the others before screen change, not here, and then we shouldn't need the if here